gazelle_haskell_modules/rule_generation.go

```diff
@@ -265,21 +267,60 @@ func concatRuleInfos(xs [][]*RuleInfo) []*RuleInfo {
 	return ys
 }
 
-// Collects the source files referenced in the given expression
-func srcsFromRule(pkgRoot string, expr build.Expr) ([]string, error) {
-	srcs, err := getSources(expr)
-	if err != nil {
-		return nil, err
+// Find either
+// 0. The files listed in the srcs attribute of a rule
+// 1. All the files (recursively) in directories specified in a "gazelle_haskell_modules:srcs:"
+// If the "srcs" attribute is present, 1. is ignored.
+func getSrcs(pkgRoot string, r *rule.Rule) ([]string, error) {
+	hasSrcs := false
+	for _, attrKey := range r.AttrKeys() {
+		if attrKey == "srcs" {
+			hasSrcs = true
+			break
+		}
 	}
 
-	xs := make([]string, len(srcs))
-	i := 0
-	for f, _ := range srcs {
-		xs[i] = path.Join(pkgRoot, f)
-		i++
-	}
+	if hasSrcs {
+		srcs, err := getSources(r.Attr("srcs"))
 
-	return xs, nil
+		if err != nil {
+			return nil, err
+		}
+
+		xs := make([]string, len(srcs))
+		i := 0
+		for f, _ := range srcs {
+			xs[i] = path.Join(pkgRoot, f)
+			i++
+		}
+
+		return xs, nil
+	} else {
+		srcDirs, err := getSrcDirsFromRuleDirective(r)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		if len(srcDirs) == 0 {
+			// TODO not sure about this, but it's the same as the err case in the previous branch
+			// which works fine..
+			return nil, nil
+		}
+
+		srcs, err := getSourcesRecursivelyFromDirs(pkgRoot, srcDirs)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		xs := make([]string, len(srcs))
```

Comment:
This construction is duplicated withing this function. BTW isn't the underscored value the index (so that you can drop this `i` counter and `i++`)?

Reply:
Good point - I copied this from an existing piece of code, so this might be a good point to try to improve both.

Reply:
Thanks, used the more readable version in the other place where this was present.